metric "diff_decorator_retry_count" was not initialized

### Steps to reproduce

Install helm chart

### Expected behavior

start normally

### Actual behavior

```
panic: metric "diff_decorator_retry_count" was not initialized
        panic: metric "diff_decorator" was not initialized
```

### Kelemetry version

0.2.2

### Environment

Install through helm chart

--- a/charts/kelemetry/templates/_helpers.yaml
+++ b/charts/kelemetry/templates/_helpers.yaml
@@ -223,6 +223,16 @@ diff-cache-patch-ttl: {{toJson .Values.informers.diff.persistDuration.patch}}
 diff-cache-snapshot-ttl: {{toJson .Values.informers.diff.persistDuration.snapshot}}
 {{- end }}
 
+{{- define "kelemetry.diff-decorator-options" }}
+{{- include "kelemetry.diff-decorator-options-raw" . | include "kelemetry.yaml-to-args" }}
+{{- end }}
+{{- define "kelemetry.diff-decorator-options-raw" }}
+diff-decorator-enable: {{ .Values.consumer.diff.enable | toJson }}
+diff-decorator-fetch-backoff: {{ .Values.consumer.diff.backoff | toJson }}
+diff-decorator-fetch-event-timeout: {{ .Values.consumer.diff.fetchEventTimeout | toJson }}
+diff-decorator-fetch-total-timeout: {{ .Values.consumer.diff.fetchTotalTimeout | toJson }}
+{{- end }}
+
 {{- define "kelemetry.event-informer-options" }}
 {{- include "kelemetry.event-informer-options-raw" . | include "kelemetry.yaml-to-args" }}
 {{- end }}
@@ -246,11 +256,6 @@ event-informer-worker-count: {{toJson .Values.informers.diff.workerCount}}
 audit-consumer-enable: true
 audit-consumer-filter-cluster-name: "" # TODO: review whether this option should be set
 audit-consumer-group-failures: false # TODO: add option to enable this when tf plugins support it
-
-diff-decorator-enable: {{ .Values.consumer.diff.enable | toJson }}
-diff-decorator-fetch-backoff: {{ .Values.consumer.diff.backoff | toJson }}
-diff-decorator-fetch-event-timeout: {{ .Values.consumer.diff.fetchEventTimeout | toJson }}
-diff-decorator-fetch-total-timeout: {{ .Values.consumer.diff.fetchTotalTimeout | toJson }}
 
 {{- if .Values.consumer.source.type | eq "webhook" }}
 audit-webhook-enable: true

--- a/charts/kelemetry/templates/_helpers.yaml
+++ b/charts/kelemetry/templates/_helpers.yaml
@@ -247,6 +247,11 @@ audit-consumer-enable: true
 audit-consumer-filter-cluster-name: "" # TODO: review whether this option should be set
 audit-consumer-group-failures: false # TODO: add option to enable this when tf plugins support it
 
+diff-decorator-enable: {{ .Values.consumer.diff.enable | toJson }}
+diff-decorator-fetch-backoff: {{ .Values.consumer.diff.backoff | toJson }}
+diff-decorator-fetch-event-timeout: {{ .Values.consumer.diff.fetchEventTimeout | toJson }}
+diff-decorator-fetch-total-timeout: {{ .Values.consumer.diff.fetchTotalTimeout | toJson }}
+
 {{- if .Values.consumer.source.type | eq "webhook" }}
 audit-webhook-enable: true
 audit-producer-enable: true

--- a/charts/kelemetry/templates/consumer.deployment.yaml
+++ b/charts/kelemetry/templates/consumer.deployment.yaml
@@ -33,6 +33,10 @@ spec:
             {{ include "kelemetry.logging-options" .Values.consumer }}
             {{ include "kelemetry.kube-options" .Values.consumer }}
             {{ include "kelemetry.audit-options" . }}
+            {{- if .Values.informers.diff.enable }}
+            {{ include "kelemetry.diff-cache-options" . }}
+            {{ include "kelemetry.diff-decorator-options" . }}
+            {{- end }}
           ]
           ports: [
             {{- if .Values.consumer.pprof }}

--- a/charts/kelemetry/values.yaml
+++ b/charts/kelemetry/values.yaml
@@ -290,8 +290,7 @@ storageBackend:
 aggregator:
   globalTags:
     # Tags applied to all object spans (i.e. not actual events, just the parent placeholder)
-    pseudoSpan:
-      cluster: foo # one recommended use is to write the cluster name here.
+    pseudoSpan: {}
     # Tags applied to all actual spans (e.g. events, audit logs)
     eventSpan: {}
 

--- a/charts/kelemetry/values.yaml
+++ b/charts/kelemetry/values.yaml
@@ -139,6 +139,24 @@ consumer:
         otherConfig: {}
           # clusterIP: xxx, externalIP: xxx, etc.
 
+  # Decorate audit logs with diff info from the diff cache.
+  diff:
+    enable: true
+    # The backoff duration between attempts to fetch diff from the diff cache.
+    backoff: 1s
+    # If this duration has elapsed since kube-apiserver sends the ResponseComplete audit stage
+    # but the diff cache still returns NotFound for the requested diff,
+    # consumer gives up retrying to fetch diff for this audit event.
+    # Consumer always tries to fetch at least once even if this duration has elapsed.
+    fetchEventTimeout: 15s
+    # The total duration that consumer takes to retry and fetch diff for each audit event.
+    # Comparing fetchEventTimeout vs fetchTotalTimeout:
+    # fetchEventTimeout is non-accumulative, e.g. if consumer is lagging behind for more than fetchEventTimeout,
+    # the consumer give up after the first attempt;
+    # meanwhile, fetchTotalTimeout is accumulative, e.g. in the worst case,
+    # the consumer takes fetchTotalTimeout to process each audit event before processing the next one.
+    fetchTotalTimeout: 10s
+
   # The audit consumer is primarily CPU-bound.
   resources: {}
     # limits:

--- a/pkg/diff/decorator/decorator.go
+++ b/pkg/diff/decorator/decorator.go
@@ -233,7 +233,10 @@ func (decorator *decorator) tryDecorate(
 			event.SetTag("newResourceVersion", *newRv)
 		}
 
-		logger = logger.WithField("oldRv", oldRv).WithField("newRv", newRv)
+		logger = logger.WithField("oldRv", oldRv)
+		if newRv != nil {
+			logger = logger.WithField("newRv", *newRv)
+		}
 
 		tryOnce = func(ctx context.Context) (bool, error) {
 			return decorator.tryUpdateOnce(ctx, object, oldRv, newRv, event, message)

--- a/pkg/manager/interface.go
+++ b/pkg/manager/interface.go
@@ -412,10 +412,14 @@ func (*List[T]) listFactory(manager *Manager) ComponentFactory {
 func (list *List[T]) listRemoveImpl(obj Component) {
 	newList := []T{}
 	for _, impl := range list.Impls {
-		if list.Indexed != nil && any(impl) == obj {
-			delete(list.Indexed, any(impl).(IndexedListImpl).ListIndex())
+		if any(impl) == obj {
+			if list.Indexed != nil {
+				delete(list.Indexed, any(impl).(IndexedListImpl).ListIndex())
+			}
+
 			continue
 		}
+
 		newList = append(newList, impl)
 	}
 

